File: app/src/main/java/org/szvsszke/vitezlo2018/MapFragment.java
package org.szvsszke.vitezlo2018;

import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;

import com.google.android.gms.maps.MapView;

import org.szvsszke.vitezlo2018.map.MapDecorator;
import org.szvsszke.vitezlo2018.map.MapPreferences;
import org.szvsszke.vitezlo2018.map.data.AbstractDataCache.DataLoadedListener;
import org.szvsszke.vitezlo2018.map.data.DescriptionsCache;
import org.szvsszke.vitezlo2018.map.model.Track;
import org.szvsszke.vitezlo2018.map.model.TrackDescription;
import org.szvsszke.vitezlo2018.map.overlay.InfoBox;
import org.szvsszke.vitezlo2018.map.overlay.MapControlBox;
import org.szvsszke.vitezlo2018.map.overlay.MapControlBox.MapControlListener;
import org.szvsszke.vitezlo2018.utilities.Utilities;

import java.util.ArrayList;

/**
 * This fragment is responsible for loading data and displaying it
 * overlaying a map. 
 * */
public class MapFragment extends Fragment implements MapControlListener {

	private static final String TAG = MapFragment.class.getName();
	private static final long TEN_SECONDS = 10000;
	
	private MapView mMapView;
	private MapDecorator mMapDecorator;
	private MapPreferences mMapPrefs;

	private RelativeLayout mInfoContainer;
	private LinearLayout mControlContainer;

	private InfoBox mInfoBox;
	
	private MapControlBox mControlBox;

	private DescriptionsCache mDescriptions;

	//view variables
	private boolean isBoxExpanded = true;

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
					Bundle savedInstanceState) {		
		Log.v(TAG, "onCreate");
		View inflatedView = inflater.inflate(R.layout.fragment_map, container,
				false);
		
		mMapPrefs = new MapPreferences(
				PreferenceManager.getDefaultSharedPreferences(getActivity()));		
		mMapView = (MapView) inflatedView.findViewById(R.id.mapView);
		mMapView.onCreate(savedInstanceState);

		mMapDecorator = new MapDecorator(getActivity(), mMapView, mMapPrefs);
		
		mDescriptions = new DescriptionsCache(getActivity());

        mInfoContainer = (RelativeLayout) 
        		inflatedView.findViewById(R.id.infoBoxHolder);
        
        if (mMapPrefs.isInfoboxEnabled()) {
        	setupInfobox(inflater);
        }
        else {
        	mInfoContainer.setVisibility(View.GONE);
        }
        
        showTrackInfo();
        
		mControlContainer = (LinearLayout) inflatedView.findViewById(
				R.id.controlBoxHolder);
		if (mMapPrefs.isControlBoxEnabled()) {			
			// setup control box
			mControlBox = new MapControlBox(mMapPrefs);	
			mControlBox.onCreateView(inflater, mControlContainer);
			mControlBox.setListener(this);		


			mControlBox.enableUserHikeButton(false);

		}
		else {
			mControlContainer.setVisibility(View.GONE);
		}
		return inflatedView;
	}

	@Override
	public void onPause() {
		Log.d(TAG, "onPause");
		super.onPause();		
		mMapView.onPause();
	}
	
	@Override
	public void onResume() {
		Log.d(TAG, "onResume");
		super.onResume();
		mMapView.onResume();
		mMapDecorator.decorate(mDescriptions.getDescription(
				mMapPrefs.getSelectedTrackIndex()));
	}
	
	@Override
	public void onDestroy() {
		Log.d(TAG, "onDestroy");
		super.onDestroy();
		mMapView.onDestroy();
		mMapPrefs.setInfoboxLocked(mInfoBox.isSpinnerLocked());
		mMapPrefs.saveMapPreferences();
	};
	
	@Override
	public void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		mMapView.onSaveInstanceState(outState);
	}
	
	@Override
	public void onLowMemory() {
		super.onLowMemory();
		mMapView.onLowMemory();
	}
	
	private void setCurrentTrack(final int trackNr) {
		Log.d(TAG, "setCurrentTrack");
		mMapPrefs.setSelectedTrackIndex(trackNr);
		if (mDescriptions.acquireData() != null) {
			mMapDecorator.decorate(mDescriptions.getDescription(trackNr));
			getActivity().setTitle(mDescriptions.getDescription(trackNr).getName());
			mInfoBox.setTitle(mDescriptions.getDescription(trackNr).getName());
		}
		else {
			mDescriptions.setDataLoadedListener(
					new DataLoadedListener<ArrayList<TrackDescription>>() {
				@Override
				public void onDataLoaded(ArrayList<TrackDescription> loaded) {
					mMapDecorator.decorate(mDescriptions.getDescription(trackNr));
					getActivity().setTitle(
							mDescriptions.getDescription(trackNr).getName());
					mInfoBox.setTitle(mDescriptions.getDescription(trackNr).getName());
				}
			});
		}
	}
	
	private void setupInfobox(LayoutInflater inflater) {
		Log.d(TAG, "setupInfoBox");
		mInfoBox = new InfoBox(getActivity());
	    mInfoBox.onCreateView(inflater, mInfoContainer);  
	    OnClickListener expandCollapse = new OnClickListener() {
	    
	    	@Override
	    	public void onClick(View v) {
	    		isBoxExpanded = !isBoxExpanded;
				mInfoBox.expandInfoBox(isBoxExpanded);									
			}
		};
		mInfoBox.setOnClickListenerForContainer(expandCollapse);
		mInfoBox.lockSpinner(mMapPrefs.isInfoboxLocked());
	}

	private void setTrackSpinner() {
		Log.d(TAG, "setTrackSpinner");
		if (mDescriptions.acquireData() != null) {
			Log.v(TAG, "description data not null");
			mInfoBox.setupSpinner(mDescriptions.getNames(), 
				new TrackSpinnerListener(), 
				mMapPrefs.getSelectedTrackIndex());
		}
		else {
			Log.v(TAG, "description data null, set listener");
			mDescriptions.setDataLoadedListener( 
				new DataLoadedListener<ArrayList<TrackDescription>>() {

					@Override
					public void onDataLoaded(
							ArrayList<TrackDescription> loaded) {
						Log.v(TAG, "mDescriptions onDataLoaded");
						mInfoBox.setupSpinner(mDescriptions.getNames(), 
								new TrackSpinnerListener(), 
								mMapPrefs.getSelectedTrackIndex());
					}
			});
		}
	}

	@Override
	public void displayPreferenceChanged() {
		Log.d(TAG, "displayPreferenceChanged");
		mMapDecorator.decorate(mDescriptions.getDescription(
				mMapPrefs.getSelectedTrackIndex()));

		showTrackInfo();
		mMapDecorator.removeUserPath();
	}
	
	private void showTrackInfo() {
		Log.d(TAG, "showTrackInfo");
		setTrackSpinner();
		updateTrackInfo();
	}

	private void updateTrackInfo() {
		Log.d(TAG, "updateTrackInfo");
		//setCurrentTrack(mMapPrefs.getSelectedTrackIndex());
		TrackDescription desc = mDescriptions.getDescription(
				mMapPrefs.getSelectedTrackIndex());
		
		if(desc!= null) {
			mInfoBox.addItems(desc.getName(), 
				getResources().getStringArray(R.array.hike_info),
				desc.getPublicData());
		} else {
			mDescriptions.setDataLoadedListener(
					new DataLoadedListener<ArrayList<TrackDescription>>() {

						@Override
						public void onDataLoaded(
								ArrayList<TrackDescription> loaded) {
							TrackDescription desc = mDescriptions.getDescription(
									mMapPrefs.getSelectedTrackIndex());
							mInfoBox.addItems(desc.getName(), 
									getResources().getStringArray(R.array.hike_info),
									desc.getPublicData());
						}
			});
		}
	}

	private class TrackSpinnerListener implements OnItemSelectedListener {

		@Override
		public void onItemSelected(AdapterView<?> parent, View view,
				int position, long id) {
			Log.d(TAG, "onItemSelected " + position);
			setCurrentTrack(position);
		}

		@Override
		public void onNothingSelected(AdapterView<?> parent) {}		
	}

}

File: app/src/main/java/org/szvsszke/vitezlo2018/MapFragment.kt
package org.szvsszke.vitezlo2018

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.gms.maps.model.LatLng
import com.xwray.groupie.ExpandableGroup
import com.xwray.groupie.GroupAdapter
import com.xwray.groupie.kotlinandroidextensions.GroupieViewHolder
import kotlinx.android.synthetic.main.fragment_map_re.*
import kotlinx.android.synthetic.main.info_box_view.*
import org.szvsszke.vitezlo2018.domain.entity.Description
import org.szvsszke.vitezlo2018.domain.entity.Point
import org.szvsszke.vitezlo2018.domain.preferences.InfoBoxStatus
import org.szvsszke.vitezlo2018.map.MapDecorator
import org.szvsszke.vitezlo2018.presentation.listitems.InfoBoxItem
import org.szvsszke.vitezlo2018.presentation.map.MapViewModel
import org.szvsszke.vitezlo2018.presentation.listitems.SimpleExpandableItem
import org.szvsszke.vitezlo2018.presentation.map.SpinnerListener
import org.szvsszke.vitezlo2018.usecase.*
import timber.log.Timber
import javax.inject.Inject

/**
 * This fragment is responsible for loading data and displaying it
 * overlaying a map.
 * TODO refactor properly once spaghetti is sorted out
 */
class MapFragment : Fragment() {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    @Inject
    lateinit var mapDecorator: MapDecorator

    @Inject
    lateinit var findBounds: FindBounds

    private lateinit var viewModel: MapViewModel

    private val groupAdapter = GroupAdapter<GroupieViewHolder>()
    private var expandableGroup = ExpandableGroup(SimpleExpandableItem(R.string.description))

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        App.getComponent().inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_map_re, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mapView.onCreate(savedInstanceState)

        recycler_infoBox.adapter = groupAdapter
        recycler_infoBox.layoutManager = LinearLayoutManager(context)

        groupAdapter.clear()
        groupAdapter.add(expandableGroup)

        mapDecorator.init(activity, mapView) // TODO

        viewModel = ViewModelProvider(this, viewModelFactory)
                    .get(MapViewModel::class.java)

        imageView_mapTypeSwitch.setOnClickListener { mapDecorator.switchMapType() }
        imageView_infoBoxCollapse.apply {
            setOnClickListener {
                val oldInfoStatus = viewModel.getInfoBoxStatus()
                val newInfoBoxStatus = oldInfoStatus.copy(isExtended = !oldInfoStatus.isExtended)
                viewModel.saveInfoBoxStatus(newInfoBoxStatus)
                expandableGroup.onToggleExpanded()
                if (newInfoBoxStatus.isExtended) {
                    setImageDrawable(ContextCompat.getDrawable(view.context, R.drawable.ic_action_collapse))
                } else {
                    setImageDrawable(ContextCompat.getDrawable(view.context, R.drawable.ic_action_expand))
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        mapView.onResume()
        mapDecorator.mapStatus = viewModel.getMapStatus()
        viewModel.getDescriptions().observe(this,
                Observer { result ->
                    when(result) {
                        is DescriptionsState.Data -> onDescriptionsReady(
                                result.descriptions, viewModel.getInfoBoxStatus())
                        else -> Timber.e("Could not get descriptions")
                    }
                })
        showSights()
        showTouristPaths()
    }

    override fun onPause() {
        viewModel.saveMapStatus(mapDecorator.mapStatus)
        mapView.onPause()
        super.onPause()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        mapView.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun onLowMemory() {
        mapView.onLowMemory()
        super.onLowMemory()
    }
    private fun onDescriptionsReady(descriptions: List<Description>,
                                    infoBoxStatus: InfoBoxStatus) {
        Timber.v(descriptions.toString())
        Timber.v(infoBoxStatus.toString())

        val hikeNames = descriptions.map { it.name }
        setupTrackSpinner(hikeNames, infoBoxStatus, descriptions)

        showCheckpoint(descriptions[infoBoxStatus.selectedTrackIndex])
        showTrack(descriptions[infoBoxStatus.selectedTrackIndex].routeFileName)
    }

    private fun setupTrackSpinner(hikeNames: List<String>, infoBoxStatus: InfoBoxStatus,
                                  descriptions: List<Description>) {
        context?.let { context ->
            val adapter = ArrayAdapter<String>(context, R.layout.info_box_spinner_item,
                    R.id.textViewTrackName, hikeNames)

            spinner_infoBoxHikeSpinner.apply {
                setAdapter(adapter)
                setSelection(infoBoxStatus.selectedTrackIndex)
                onItemSelectedListener = SpinnerListener { position ->
                    updateContent(descriptions[position])
                }
            }
        }
        showSpinnerOrTitle(hikeNames, infoBoxStatus)
        imageView_infoBoxLock.setOnClickListener {
            val oldInfoStatus = viewModel.getInfoBoxStatus()
            val newInfoBoxStatus = oldInfoStatus.copy(isLocked = !oldInfoStatus.isLocked)
            viewModel.saveInfoBoxStatus(newInfoBoxStatus)
            showSpinnerOrTitle(hikeNames, newInfoBoxStatus)
        }
    }

    private fun updateContent(description: Description) {
        showTrack(description.routeFileName)
        showCheckpoint(description)
        updateDescription(description)
    }

    private fun showSpinnerOrTitle(hikeNames: List<String>, infoBoxStatus: InfoBoxStatus) {
        if(infoBoxStatus.isLocked) {
            textView_infoBoxTitle.visibility = View.VISIBLE
            textView_infoBoxTitle.text = hikeNames[infoBoxStatus.selectedTrackIndex]
            spinner_infoBoxHikeSpinner.visibility = View.GONE
            context?.let { context ->
                imageView_infoBoxLock.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_menu_lock))
            }
        } else {
            textView_infoBoxTitle.visibility = View.GONE
            spinner_infoBoxHikeSpinner.visibility = View.VISIBLE
            spinner_infoBoxHikeSpinner.setSelection(infoBoxStatus.selectedTrackIndex)
            context?.let { context ->
                imageView_infoBoxLock.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_menu_unlock))
            }
        }
    }

    private fun updateDescription(description: Description) {
        groupAdapter.remove(expandableGroup)
        expandableGroup = ExpandableGroup(SimpleExpandableItem(R.string.description))
        expandableGroup.addAll(listOf(
                InfoBoxItem(R.string.length, description.length),
                InfoBoxItem(R.string.leveltime, description.levelTime),
                InfoBoxItem(R.string.starting, description.starting),
                InfoBoxItem(R.string.fee, description.entryFee)
        ))
        if(description.other.isNotBlank()) {
            expandableGroup.add(InfoBoxItem(R.string.other, description.other))
        }
        groupAdapter.add(expandableGroup)
        groupAdapter.notifyDataSetChanged()
    }

    private fun showTrack(trackName: String) {
        viewModel.getTrack(trackName).observe(this,
                Observer { result ->
                    when(result) {
                        is TrackState.Data -> processTrack(result.data.points)
                        else -> Timber.e("Could not get track")
                    }
                })
    }

    private fun showCheckpoint(description: Description) {
        viewModel.getCheckpoints(description.checkPointIDs).observe(this,
                Observer<CheckpointState> { result ->
                    when (result) {
                        is CheckpointState.Data -> mapDecorator.markCheckpoints(result.data)
                        is CheckpointState.Disabled -> mapDecorator.hideCheckpoints()
                        else -> Timber.e("Could not get checkpoints")
                    }
                })
    }

    private fun showSights() {
        viewModel.getSights().observe(this,
                Observer { result ->
                    when(result) {
                        is SightsState.Data -> mapDecorator.markSights(result.data)
                        is SightsState.Disabled -> mapDecorator.hideSights()
                        else -> Timber.e("Could not get sights")
                    }
                })
    }

    private fun showTouristPaths() {
        viewModel.getTouristPath().observe(this,
                Observer { result ->
                    when(result) {
                        is TouristPathState.Data -> mapDecorator.displayTouristPaths(result.data)
                        is TouristPathState.Disabled -> mapDecorator.hideTouristPaths()
                        else -> Timber.e("Could not get tourist paths")
                    }
                })
    }

    // TODO move to a utility class
    private fun processTrack(track: List<Point>) {
        val bounds = findBounds(track)
        val start = LatLng(bounds.first.latitude, bounds.first.longitude)
        val end = LatLng(bounds.second.latitude, bounds.second.longitude)
        mapDecorator.markTrack(track.map { LatLng(it.latitude, it.longitude) }, start, end)
    }



}
